src/storage/observer_worker.rs

```diff
@@ -99,56 +97,41 @@ where
         }
     }
 
-    async fn try_update(&self) -> Result<()> {
-        trace!("try update active blob");
-        let inner_cloned = self.inner.clone();
-        if let Some(inner) = active_blob_check(inner_cloned).await? {
-            update_active_blob(inner.clone()).await?;
-            inner.try_dump_old_blob_indexes(self.dump_sem.clone()).await;
-        }
-        Ok(())
-    }
-}
+    async fn try_update_active_blob(&self) -> Result<bool> {
+        let config_max_size = self
+            .inner
+            .config
+            .max_blob_size()
+            .ok_or_else(|| Error::from(ErrorKind::Uninitialized))?;
+        let config_max_count = self
+            .inner
+            .config
+            .max_data_in_blob()
+            .ok_or_else(|| Error::from(ErrorKind::Uninitialized))?;
 
-async fn active_blob_check<K>(inner: Inner<K>) -> Result<Option<Inner<K>>>
-where
-    for<'a> K: Key<'a> + 'static,
-{
-    let (active_size, active_count) = {
-        trace!("await for lock");
-        let safe_locked = inner.safe.read().await;
-        trace!("lock acquired");
-        if let Some(active_blob) = safe_locked.active_blob.as_ref() {
-            (active_blob.file_size(), active_blob.records_count() as u64)
-        } else {
-            // if active blob doesn't exists, it doesn't need to be updated
-            return Ok(None);
+        {
+            let read = self.inner.safe.read().await;
+            let active_blob = read.active_blob.as_ref();
+            if let Some(active_blob) = active_blob {
+                if active_blob.file_size() < config_max_size
+                    && (active_blob.records_count() as u64) < config_max_count
+                {
+                    return Ok(false);
+                }
+            }
         }
-    };
-    trace!("lock released");
-    let config_max_size = inner
-        .config
-        .max_blob_size()
-        .ok_or_else(|| Error::from(ErrorKind::Uninitialized))?;
-    let config_max_count = inner
-        .config
-        .max_data_in_blob()
-        .ok_or_else(|| Error::from(ErrorKind::Uninitialized))?;
-    if active_size as u64 > config_max_size || active_count >= config_max_count {
-        Ok(Some(inner))
-    } else {
-        Ok(None)
+        update_active_blob(&self.inner).await.map(|_| true)

```

Comment:
This update should be done inside `write` lock with limit checks (as it was before the last commit). My previous comment was only about the logic after all the checks

Reply:
Why? It can't be updated elsewhere, so there is no need to keep write lock

Reply:
Here is the public method, that updates acitve blob directly (without observer): https://github.com/qoollo/pearl/blob/master/src/storage/core.rs#L178

Reply:
Ah, since blob can be missing now we have synchronous operations... You're right, write lock should be held.